### ISSUE TITLE
Type information not serialized for map values

### ISSUE THREAD

**Issue by gianm** [FasterXML/jackson-databind#3583]
**Describe the bug**
When serializing `Map<String, I>`, where `I` is an interface with `@JsonTypeInfo`, the type information for instances of `I` is omitted. However, when serializing the same map as entries one-by-one, the type information is included, which shows that it _could_ be included in principle for serialization of the Map itself.

I wrote this as a bug report, but I'm wondering if it's really a bug, or if it's intentional behavior for some reason.

**Version information**
2.13.3

**To Reproduce**
JUnit test here: https://gist.github.com/gianm/8e11ab6ce41a17ceff83ade37c02d112

In this test file, `writeAsMap` fails and `writeAsEntries` passes.

**Comment by gianm** [comment]
Btw, in looking into this a bit, I've learned that SerializerProvider has `findValueSerializer` (does not write type info) and `findTypedValueSerializer` (does write type info). Main difference between the two cases in the [gist I posted](https://gist.github.com/gianm/8e11ab6ce41a17ceff83ade37c02d112) seems to be that the MapSerializer uses the serializer from `findValueSerializer` for values, whereas when writing entries one-by-one, the `objectMapper.writeValue(jg, entry.getValue())` call ultimately ends up using the serializer from `findTypedValueSerializer`.

I'm not sure if there is a reason for this, or if MapSerializer could be updated to use the typed serializer for values too.

**Comment by cowtowncoder** [comment]
If this is root-level `Map`, the problem is Java Type Erasure which makes it impossible to know the proper base type of `Map` values; and it's basically only seen as `Map<?,?>`.
And while contents may still be serialized as the actual runtime class is available, it is not possible to determine the Base Type necessary to determine if Polymorphic Types are needed.
If so, this is basically working as expected, unfortunately.

There are couple of usual work-arounds. Either:

1. Use a throw-away type where type becomes non-generic -- like `public class StringMyValueMap extends HashMap<String, MyValue> { }` -- and pass instances of that. Type information will be accessible via super-type (as the class is no longer generic)
2. Specify type for `ObjectWriter` (something like `ObjectMapper.writer().forType(new TypeReference<Map<String, MyValue>>() { })`): this will indicate actual type as well.
3. Use a wrapper object (non-generic) that only has generic `Map` property -- type information is retained here as well.

So fundamentally I always recommend avoiding use of generic types (`Map`, `Collection`) as Root Values since that is the one place where Java Type Erasure messes things up.

**Comment by FrankChen021** [comment]
Hello @cowtowncoder

We know that type info on generic class will be erased, but we think that the problem is, if this is the root cause, why does the customized serialization work?

Since the poly type info is annotated on the class of values in the `Map`, I think the type info does not lost.

I dived into the current implementation of `MapSerializer`, 
```java
    public void serialize(Map<?,?> value, JsonGenerator gen, SerializerProvider provider)
        throws IOException
    {
        gen.writeStartObject(value);
        serializeWithoutTypeInfo(value, gen, provider);
        gen.writeEndObject();
    }
```

As we can see from the function name `serializeWithoutTypeInfo`, It serializes the key and values without type info intentionally(Why?) 

This method then calls following method to serialize keys and values.

```java
public void serializeFields(Map<?,?> value, JsonGenerator gen, SerializerProvider provider)
        throws IOException
    {
       ...
        try {
            for (Map.Entry<?,?> entry : value.entrySet()) {
                Object valueElem = entry.getValue();
                ...
                
                JsonSerializer<Object> serializer = _valueSerializer;
                if (serializer == null) {
                    serializer = _findSerializer(provider, valueElem);
                }
                serializer.serialize(valueElem, gen, provider);
            }
        } catch (Exception e) { // Add reference information
            wrapAndThrow(provider, e, value, String.valueOf(keyElem));
        }
    }
```

ANd the `_findSerializer` method will look up a serializer which DOES NOT write the type info into target string.

However, If we register a customised Map serializer, and replace `_findSerializer` by calling `writeObject`, type info IS included in the final output.

```java
    static class OurMapSerializer extends StdSerializer<Map> {
        public OurMapSerializer() {
            super(Map.class);
        }

        @Override
        public void serialize(Map value, JsonGenerator gen, SerializerProvider provider) throws IOException {
            gen.writeStartObject();
            for (Object entry : value.entrySet()) {
                Object k = ((Map.Entry<?, ?>)entry).getKey();
                Object v = ((Map.Entry<?, ?>)entry).getValue();

                provider.findKeySerializer(k.getClass(), null).serialize(k, gen, provider);
                gen.writeObject(v); //<===writeObject instead of using _findSerializer to get serializer first
            }
            gen.writeEndObject();
        }
    }
```

So, I think the problem is not that the current implementation of `MapSerializer` in jackson does not know the type info, but does not use type serializer to serialize values in `Map`.

**Comment by cowtowncoder** [comment]
Correct: information on serialization differs from that of deserialization. This can cause issues when serializer/deserializer disagree on base type; and from that whether `Type[De]Serializer` (handler for polymorphic type id) is to be used or not.

Use of `JsonGenerator.writeObject(...)` is usually wrong way to do things, fwtw, since it will lose context. There are methods in `SerializerProvider`, and `findSerializer()` (etc) are the ways to go.
Although the best approach is to try to resolve key and value serializers in `createContextual()` method (of `ContextualSerializer` interface) if type is known from type declaration.

As to why "serializeWithoutTypeInfo()" -- this is because serializer knows type information is NOT to be written in this particular case. It is just an optimization.

I wish I had time to look into this deeper but as of now I don't have time to.

**Comment by cowtowncoder** [comment]
Ok, one more thing: `JsonGenerator`'s `writeObject()` will essentially look up Root Value Serializer for type `value.getClass()`. Since you are handing specific value instance, its full type is known and used for determining "base type" for write Type Id (if polymorphic handling annotations found).
This is different from `Map` being handed as Root Value, which is also used for values.
That type is likely lost and mostly assumed `Object`.
But it is essential for deserialization that the base type of ALL values is the same; so per-value lookup is potentially wrong and resulting in deserialization failing (since deserializer will use same base type and then `TypeDeserializer` (if any); and if actual `TypeSerializer` used on serialization varies there's mismatch).

That's a long way of saying that custom serializer's handling is incompatible with the way deserialization works.

But if it appears to work for your case (maybe you don't deserialize anything?), you may want to just use that. It is unlikely that default handling of `Map[De]Serializer` in this area will be changed.

**Comment by gianm** [comment]
> But if it appears to work for your case (maybe you don't deserialize anything?), you may want to just use that. It is unlikely that default handling of Map[De]Serializer in this area will be changed.

You're right. In the cases where this came up for us, in fact we do not deserialize the JSON in our own code. Where this came up, the serialization is done as part of logging or reporting, and it's basically write-only. Our usage is something like this:

1. Code adds a bunch of stuff to `Map<String, Object>` and passes it to a logger.
2. Logger serializes the Map (using Jackson) and writes it to log files. The logger does not know what the stuff is.
3. Our own software does not parse the log files, but other downstream consumers do, and they "know" what keys are expected for certain kinds of logged events. For example, they would read `query` out of the map for certain log events and deserialize it as the `Query.class` interface (ignoring other kinds of log events).

Unfortunately due to the wide variety of stuff that might be logged, it's not feasible for us to replace the `Map<String, Object>` with a concrete class.

It sounds like, for this kind of usage, builtin Jackson MapSerializer wouldn't be changed so a custom Map serializer is the way to go?

**Comment by cowtowncoder** [comment]
@gianm Yes, correct: with custom implementation you are free to apply whatever logic works best for your use case.
The default `MapSerializer` / `MapDeserializer` combo is limited by a few factors:

1. Trying to ensure round-trip capability (deseralizer having less information)
2. Related to (1) serializer must operate on nominal (declared) type of `Map` with generics, instead of looking at actual values
3. For general consistency the base type of all Map values is considered the same, allowing `TypeSerializer` to be constructed once before use, instead of trying to determine it on per-value basis (also avoiding non-trivial performance overhead).

Custom handlers, howevers,  can take into account specific case you have, with possibly implicit limits and guarantees.
